### PR TITLE
Adding actions to generate paper results

### PR DESCRIPTION
**Pull request by Machiry**
* For the non `-alltypes` variant, run sub-variants with different solving strategies: `-only-g-sol`, `-only-l-sol`, or neither.
* Upload stats as artifacts (with 5 retention days)

Latest, we created a new workflow `exhaustive.yaml` different from the current workflow and helps us keep things separate from general testing.

### CODE REVIEW

**Comment by mattmccutchen-cci** [comment]
I took a quick look at this PR, and I think I understand at a high level what it is trying to achieve.  However, I see several small problems and inconsistencies, and I think the Python implementation could use a lot of cleanup.  I'd like to address both of those things if you think we can afford the time, @mwhicks1.

We should also be mindful of the interaction between this and #16.  For one thing, they'll need to be merged, though #16 is small enough that that should be manageable.  Also, this and #16 will each approximately double the number of benchmark jobs and their total running time.  Currently, we have a fixed 3 jobs totaling ~13 minutes for the 3C build and regression tests, plus 14 jobs (7 benchmarks * (no `-alltypes` + `-alltypes`)) totaling ~6 minutes for the benchmarks.  With either this PR _or_ #16, we'll be at 31 jobs totaling ~25 minutes.  With _both_ PRs, we'll be at 59 jobs totaling ~37 minutes.  The running time may not be a big problem (we already have some ideas to shave more time off in #8), but it will become unwieldy to browse that many jobs in the GitHub UI.  We should think about reducing the number of combinations we run.

**Comment by mattmccutchen-cci** [comment]
@Machiry Did you intend to add this PR to [a milestone](https://github.com/correctcomputation/actions/milestone/1)?  If so, can you please give it a more meaningful name and description?

**Comment by Machiry** [comment]
> Why are these "fixes" ? It looks like you just changed the indentation? And, why was that necessary?

It was necessary. Making sure that the yaml formats properly and actions run in the correct order.

**Comment by Machiry** [comment]
> I took a quick look at this PR, and I think I understand at a high level what it is trying to achieve. However, I see several small problems and inconsistencies, and I think the Python implementation could use a lot of cleanup. I'd like to address both of those things if you think we can afford the time, @mwhicks1.
> 
> We should also be mindful of the interaction between this and #16. For one thing, they'll need to be merged, though #16 is small enough that that should be manageable. Also, this and #16 will each approximately double the number of benchmark jobs and their total running time (not counting the artifact upload; I'm not familiar with how long that takes). Currently, we have a fixed 3 jobs totaling ~13 minutes for the 3C build and regression tests, plus 14 jobs (7 benchmarks * (no `-alltypes` + `-alltypes`)) totaling ~6 minutes for the benchmarks. With either this PR _or_ #16, we'll be at 31 jobs totaling ~25 minutes. With _both_ PRs, we'll be at 59 jobs totaling ~37 minutes. The running time may not be a big problem (we already have some ideas to shave more time off in #8), but it will become unwieldy to browse that many jobs in the GitHub UI. We should think about reducing the number of combinations we run. (@mwhicks1 and @Machiry, did you already discuss any of this in a forum that is not visible to me?)

We can create 2 workflows: The current one (that could run on every push) and an exhaustive one with all the combinations (runs once a day).

Please feel free to change the python script and the workflow files.

**Comment by Machiry** [comment]
> @Machiry Did you intend to add this PR to [a milestone](https://github.com/correctcomputation/actions/milestone/1)? If so, can you please give it a more meaningful name and description?

Removed the other pull request.

**Comment by Machiry** [comment]
> > We can create 2 workflows: The current one (that could run on every push) and an exhaustive one with all the combinations (runs once a day).
> 
> Sounds reasonable. But I am against forking the entire `generate-workflow.py`: there is a reason why I factored out the logic into a script in the first place rather than continuing to maintain copy-pasted code in the workflow file.  It's highly likely that we'll want to make at least one change to both workflows before the paper submission (starting with #16), which would be more work if we merged this PR in its current form. Instead, you can just have `generate-workflow.py` update both files every time it runs by calling a subroutine twice with different arguments.

Agree. @mattmccutchen-cci Can you please check now!

File: generate-workflow.py [hunk 1]
```diff
@@ -323,92 +324,173 @@ def __str__(self):
         return textwrap.indent(part1 + part2, 6 * ' ')
 
 
+class ActionStep:
```
Comment (mattmccutchen-cci) [inline generate-workflow.py:327]: Use `NamedTuple` and/or factor out the duplicated code with `Step` (both if possible; if not, then we'll have to pick one).

File: generate-workflow.py [hunk 2]
```diff
@@ -323,92 +324,173 @@ def __str__(self):
...
+    workflow_name = workflow_file_configs[workflow_file][0]
+    run_configurations = workflow_file_configs[workflow_file][1]
+    generate_stats = workflow_file_configs[workflow_file][2]
+    with open(workflow_file, 'w') as out:
+        out.write(HEADER.replace('{workflow.name}', workflow_name))
+        for binfo in benchmarks:
+            for alltypes in run_configurations:
+                create_benchmark(out, binfo, alltypes, generate_stats)
+                extra_args = run_configurations[alltypes]
+                if len(extra_args) > 0:
```
Comment (mattmccutchen-cci) [inline generate-workflow.py:494]: Remove this conditional: if `len(extra_args) == 0`, the for loop does nothing anyway.

File: generate-workflow.py [hunk 3]
```diff
@@ -323,92 +324,173 @@ def __str__(self):
...
+    '.github/workflows/main.yml': ("3C benchmark tests",
+                                   regular_run_configurations, False),
+    '.github/workflows/exhaustive.yml': ("Exhaustive testing and Performance Stats",
+                                         exhaustive_run_configurations, True)
+}
+
+for workflow_file in workflow_file_configs:
+    workflow_name = workflow_file_configs[workflow_file][0]
+    run_configurations = workflow_file_configs[workflow_file][1]
+    generate_stats = workflow_file_configs[workflow_file][2]
```
Comment (mattmccutchen-cci) [inline generate-workflow.py:487]: Use `NamedTuple` instead of subscripting the tuples numerically.

File: generate-workflow.py [hunk 4]
```diff
@@ -323,92 +324,173 @@ def __str__(self):
...
+    at_flag = '--extra-3c-arg=-alltypes \\\n' if alltypes else ''
+
+    job_suffix = ''
+    for earg in extra_args:
+        if not job_suffix:
+            job_suffix = "_"
+        at_flag += '--extra-3c-arg=' + earg + ' \\\n'
+        job_suffix += earg.replace('-', '_')
+
+    at_dir = ('${{env.benchmark_conv_dir}}/' +
```
Comment (mattmccutchen-cci) [inline generate-workflow.py:362]: `at` here was short for `alltypes`.  If this variable is going to be generalized to include information not related to `alltypes`, its name should be updated.

File: generate-workflow.py [hunk 5]
```diff
@@ -199,7 +200,7 @@ class BenchmarkInfo(NamedTuple):
 # generate-workflow.py instead and re-run it. Some things in this file are
 # explained by comments in generate-workflow.py.
 
-name: 3C benchmark tests
+name: {workflow.name}
 
 on:
   # Run every day at 3:00 AM EDT (i.e., 7 am UTC)
```
Comment (mattmccutchen-cci) [inline generate-workflow.py:206]: The schedule times of the two workflows should probably be spaced well apart because of #6.

**Review by mattmccutchen-cci — CHANGES_REQUESTED** [review]
I think the design is basically right now (thanks), but the Python code could still use a bunch of cleanup.  I've commented on most (but not all) of the issues I identified.  If you're interested, you can work on them yourself, otherwise I'm happy to make the changes I want myself but I probably won't get to it until late tomorrow.

**Comment by mattmccutchen-cci** [comment]
Thanks Aravind for the fixes.  I think I'm going to take over the work for now because that will be more efficient than writing up all the changes I want in order for you to make them; I'll submit another PR so you can review my changes.

**Review by mattmccutchen-cci — CHANGES_REQUESTED** [review]
Now I want the cleanups in #18 (subject to further review by Aravind).  Either we can merge #18 into this PR (after which I'd approve this PR) or we can merge #18 into `main` directly.

**Comment by mattmccutchen-cci** [comment]
FTR, I approve this PR; it doesn't seem that GitHub will let me formally approve it or even dismiss my previous "changes requested" now that the PR is merged.

**Pull request merged**